### DISCUSSION TITLE
Should we move support from Slack to GitHub Discussions?

### DISCUSSION THREAD

**Discussion by nwalters512** [PrairieLearn/PrairieLearn#6018] [Polls]
Myself and some of the core maintainers of PrairieLearn (@mwest1066 and @mfsilva22) are considering moving a lot of technical and user support from Slack to GitHub Discussions. We'd like to lay out the rationale for this, explain what would change, and seek input from y'all in the PL community!

## Rationale

We're on the Slack free plan, which means that we quickly lose access to our chat history (things "age out" after just a few months). This makes it difficult to check if a question was asked and answered in the past, provide links to old answers, etc. By moving discussions to GitHub Discussions, every question and answer will contribute towards an indexable, searchable, permanent body of knowledge about PrairieLearn, which is something we’re really excited about!

One specific benefit of putting more discussions on GitHub is that they'll be indexed by search engines, which means folks searching for "how do I XYZ prairielearn" on Google would get meaningful results as we build a body of knowledge here. This would augment our documentation to ensure folks are getting useful results wherever they're searching.

## What would change

We'd start encouraging folks to post the following types of questions/discussions for GitHub instead of Slack:

* "How do I do X with PrairieLearn?": These would go in ["Q&A"](https://github.com/PrairieLearn/PrairieLearn/discussions/categories/q-a) category. This provides an experience like Stack Overflow where answers can be voted on and marked as accepted.
* "I'd like to request feature X": These would go in the ["Ideas"](https://github.com/PrairieLearn/PrairieLearn/discussions/categories/ideas) category. Ideas can be voted on, which will help maintainers gauge interest into different ideas. And when we're ready to implement them, we can quickly convert them from a discussion to an issue (or link another issue to track implementation).
* Questions about developing PrairieLearn ("How do I run locally?" or "How might I implement feature X?") would go in the ["Dev"](https://github.com/PrairieLearn/PrairieLearn/discussions/categories/dev) category.

In addition to questions, we'd start using the ["Announcements"](https://github.com/PrairieLearn/PrairieLearn/discussions/categories/announcements) category to announce new deploys of PrairieLearn and the features that have been released. This would take the place of the deploy announcements in the `#announcements` channel on Slack.

## What would not change

Slack still has some good use cases, like real-time support in the case of a service disruption, or communication between UIUC CBTF proctors and course staff. We also think it's a great place for PL folks to hang out and chat (looking at you, `#random`!). In general, when there's not value in a discussion being index and searchable forever, it's fine to keep it on Slack!

## Seeking feedback

Before we make this move, we wanted to get input from the PL community first to see if y'all feel that this would work well, or if you have concerns about how it might work in practice. Take a second to vote in the poll and provide any comments you have!

**Comment by stephan-koenig** [comment]
Overall this makes sense to me. The only thing to consider is if all PL users are familiar with GitHub, i.e. would it pose a more considerable barrier to expect people to participate on GitHub vs on Slack?

**Reply by mwest1066** [reply]
It's a good point that GitHub discussions don't have a very obvious workflow if you aren't logged in to GitHub. Unlike other web forums, they don't have an obvious "click here to make an account to post" button, which is unfortunate. We can try and mitigate this by telling people that they will need make a GitHub account to post questions when we link to Discussions from the docs and other places. Any other ideas for helping people who aren't familiar with GitHub?

**Reply by jonatanschroeder** [reply]
Not sure if this is that big of a deal. Slack also requires an account, and most instructors would benefit from having a Github account anyway to access their course repos.

**Comment by coatless** [comment]
My main worry outside of professors not being familiar with GitHub is that dropping everything into GitHub discussions may yield questions that can be somewhat sensitive and unintentionally expose issues. 

If the goal is to retain history, why not move over to the MS Teams license that UIUC has? This would preserve the history of chat. 

Alternatively, if there are frequent questions appearing, maybe do a "summary" write-up at the end of the week that gets posted publically?

**Reply by mfsilva22** [reply]
I don't think the intention is to drop "everything" on GutHub discussions. We agree that some sensitive topics should stay at Slack (especially the ones that are specific to the various institution using PL). 

We believe Github Discussions will be a great place for general questions "How can I create an assessment that is graded...". We get so many questions like this, and they are of interest of the entire community, and it is sad to see these discussions disappearing. For example, I have seen discussions about topics that at the time were not relevant to me, but at later time I wish I could have access to them!

This hybrid community support with Github Discussion and Slack may not be the most ideal, but so far the best solution we could come up. We are always open to suggestions!

**Reply by coatless** [reply]
@mfsilva22 I understand the growing pains and need to decrease reliance on UIUC-specific resources. However, I'm not sold that adopting GitHub discussions is the best approach even for a subset of questions. 

How about spending a small amount of time engineering a solution to the problem of persistency by enabling external logging of the Slack discussions to a web portal? 

As it stands now, we can set up a [matrix bridge](https://matrix.org/bridges/) into an IRC channel of the five different rooms of interest: `#announce`, `#pl-dev`, `#pl-help`, `#general`, and `#random`. From there, we can log messages via IRC and post them publically to a website that would allow for us to edit/sanitize logs on a 10-minute delay. As an example, the `MLPACK` library uses: [whitequark's `irclogger`](https://github.com/whitequark/irclogger) to make available their logs, c.f.: https://libera.irclog.whitequark.org/mlpack/2022-07-03

**Reply by echuber2** [reply]
Sorry to bump this thread. I've seen a few communities trying out the Matrix IRC bridge and it seems it can be spotty at times. Another thing worth mentioning might be Discord, which has a similar look and feel to IRC and permanent message history for free (at least, currently).

**Comment by ffund** [comment]
I just want to put on the record that the "feel" of these platforms and the way I use them is very different (for me, at least):

- Slack: feels like a community of educators helping each other out with assessment
- Github: feels like tech support for the platform

- Slack: feels casual and ephemeral, I am less averse to asking "dumb" questions
- Github: feels permanent (I understand that this is the point here...) and very public 
 
- Slack: I am part of multiple workspaces, including some I am not very engaged in and don't get notifications for. But I always have Slack running when I'm at work. When workspaces I'm not very engaged in have activity, I will notice eventually (because of the dot thing in the sidebar) and I will probably catch up on that activity when I have a moment.
- Github: As a casual PL user, I wouldn't subscribe to email notifications for discussions. So I'd only notice activity if I actively opened this page in my browser (which, realistically, I only would if I had an immediate question or problem myself). 

So a move to more activity on Github, less activity on Slack would definitely impact the type of engagement a hypothetical "me" first trying PL would have, and also the way I would perceive my "relationship" with the platform (user vs community member).

**Reply by nwalters512** [reply]
This is really good feedback, I definitely agree that Slack feels much more community-like.

Here's a different way of thinking about how we'd use GitHub Discussions: Slack isn't going anywhere, and folks can continue asking questions there if they're more comfortable, but we'd encourage folks to post questions and look for answers on GitHub as a first line of defense. And if they'd rather ask on Slack, once they have an answer, we can encourage them to post their question + answer on GitHub so that it can contribute to a permanent and shared knowledge base.

If you or others would rather keep up with discussions via Slack, I just created a [#pl-github-discussions](https://prairielearn.slack.com/archives/C03PXS8SMM5) Slack channel that will receive all GitHub Discussions activity!

FWIW, I think the "dumb" questions (no such thing!) are the ones that people have over and over, and it's even more important for those questions and answers to be somewhere more accessible than Slack!

**Reply by firasm** [reply]
One more thing to add: over time, as the body of knowledge builds on GitHub Discussions, one could imagine someone asking a question on PL, and then an active user responding with a link to a discussion with that question answered. 

If that Q hasn't been answered already, the user is politely encouraged to create a post here.

I think that's a very legitimate model, it preserves questions long term and allows us to use slack for ephemeral conversations. 

I see GitHub Discussions more along the lines of StackOverflow and less like a "community forum" or message board.

**Reply by nwalters512** [reply]
@firasm yes! Linking back to existing answers is exactly what I had in mind.

Semi-related to that, if we find ourselves linking to the same answer over and over, that might be a good sign that it's time to lift information up into the official documentation somehow.
